### PR TITLE
Feature/1password

### PR DESCRIPTION
**Pull request by azurekid**
image
   
   Change(s):
   - Added the following artifacts:
   ## Alert Rules
    - 1Password - Changes to firewall rules.yaml
    - 1Password - Changes to SSO configuration.yaml
    - 1Password - Disable MFA factor or type for all user accounts.yaml
    - 1Password - Log Ingestion Failure.yaml
    - 1Password - Manual account creation.yaml
    - 1Password - New service account integration created.yaml
    - 1Password - Non-privileged vault user permission change.yaml
    - 1Password - Potential insider privilege escalation via group.yaml
    - 1Password - Potential insider privilege escalation via vault.yaml
    - 1Password - Privileged vault permission change.yaml
    - 1Password - Secret extraction post vault access change by administrator.yaml
    - 1Password - Service account integration token adjustment.yaml
    - 1Password - Successful anomalous sign-in.yaml
    - 1Password - User account MFA settings changed.yaml
    - 1Password - User added to privileged group.yaml
    - 1Password - Vault export post account creation.yaml
    - 1Password - Vault export prior to account suspension or deletion.yaml
    - 1Password - Vault export.yaml
   
   ## Data Connector
   
- 1Password_API_FunctionApp.json
   
   ## Workbooks

   - 1Password.json 

   Reason for Change(s):
   - New feature for Microsoft Sentinel Content Hub

   Version Updated:
   - Yes

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - In Progress

### CODE REVIEW

**Comment by v-prasadboke** [comment]
Hello @azurekid, Thanks for raising this PR. This PR will be investigated and we will update you about the same before 23 February, 2024

**Comment by scottisloud** [comment]
Hi @azurekid looking forward to really digging into the latest work here :)

**Comment by azurekid** [comment]
Hi @v-prasadboke @v-atulyadav 

I have looked at the error in the pipeline and it seems something related to the pester test being executed. 
Based on the errors, I have gone through the `mainTemplate.json` which is the file being created by the createSolutionV3.ps1 based on the yaml files in the pull request.

What I can see is that the test fails on either empty arrays `: []` and boolean values.
These booleans are mandatory in the YAML files as they are part of the analytics rule configuration.

image

I have already did a shoutout on the Microsoft MVP community and CCP channels who could help me, but got not response yet.

This solution has been build with approval of 1Password and they really want to have their solution into the content hub as it is one of the most requested features from their customers.

**Comment by v-prasadboke** [comment]
Hello @azurekid, 

1. Please create a custom table named `OnePasswordEventLogs_CL` at location `.script/tests/KqlvalidationsTests/CustomTables`
2. Add workbook metadata to this file `Workbooks/WorkbooksMetadata.json`

**Comment by v-prasadboke** [comment]
> Hi @v-prasadboke @v-atulyadav
> 
> I have looked at the error in the pipeline and it seems something related to the pester test being executed. Based on the errors, I have gone through the `mainTemplate.json` which is the file being created by the createSolutionV3.ps1 based on the yaml files in the pull request.
> 
> What I can see is that the test fails on either empty arrays `: []` and boolean values. These booleans are mandatory in the YAML files as they are part of the analytics rule configuration.
> 
> image
> 
> I have already did a shoutout on the Microsoft MVP community and CCP channels who could help me, but got not response yet.
> 
> This solution has been build with approval of 1Password and they really want to have their solution into the content hub as it is one of the most requested features from their customers.

I'll remove it from my side @azurekid

**Comment by azurekid** [comment]
> > Hi @v-prasadboke @v-atulyadav
> > I have looked at the error in the pipeline and it seems something related to the pester test being executed. Based on the errors, I have gone through the `mainTemplate.json` which is the file being created by the createSolutionV3.ps1 based on the yaml files in the pull request.
> > What I can see is that the test fails on either empty arrays `: []` and boolean values. These booleans are mandatory in the YAML files as they are part of the analytics rule configuration.
> > image
> > I have already did a shoutout on the Microsoft MVP community and CCP channels who could help me, but got not response yet.
> > This solution has been build with approval of 1Password and they really want to have their solution into the content hub as it is one of the most requested features from their customers.
> 
> I'll remove it from my side @azurekid

Cool!,

Can you please let me know what you have updated when it's working so I can save you the work next time ;-)
Always open to learn from others.

**Comment by v-prasadboke** [comment]
> > > Hi @v-prasadboke @v-atulyadav
> > > I have looked at the error in the pipeline and it seems something related to the pester test being executed. Based on the errors, I have gone through the `mainTemplate.json` which is the file being created by the createSolutionV3.ps1 based on the yaml files in the pull request.
> > > What I can see is that the test fails on either empty arrays `: []` and boolean values. These booleans are mandatory in the YAML files as they are part of the analytics rule configuration.
> > > image
> > > I have already did a shoutout on the Microsoft MVP community and CCP channels who could help me, but got not response yet.
> > > This solution has been build with approval of 1Password and they really want to have their solution into the content hub as it is one of the most requested features from their customers.
> > 
> > 
> > I'll remove it from my side @azurekid
> 
> Cool!,
> 
> Can you please let me know what you have updated when it's working so I can save you the work next time ;-) Always open to learn from others.

You can remove the empty properties from maintemplate.

**Comment by azurekid** [comment]
> maintemplate

Yeah sure,
The only things is that with every new release we need to manually go through the `mainTemplate.json` and remove the empty array properties that have been added by the Microsoft Script.

Is this correct?

**Comment by v-prasadboke** [comment]
> > maintemplate
> 
> Yeah sure, The only things is that with every new release we need to manually go through the `mainTemplate.json` and remove the empty array properties that have been added by the Microsoft Script.
> 
> Is this correct?

yes, usually this doesnt happens. 
There must be some properties in your rules which have been kept empty. 
While repackaging arm ttk fails for the same (empty brackets) which needs to be removed from the maintemplate

**Comment by scottisloud** [comment]
Hi @v-prasadboke 

I wanted to chime in here to introduce myself. I am a Solutions Architect from 1Password and have been working with @azurekid as he has been building this Solution. 

I am wondering what the best way for me (on behalf of 1Password) to contribute to this PR so we can assist with the final polish and presentation without having to burden @azurekid with lots of little changes and enhancements? 

We are excited to continue supporting this work, and are immensely grateful to @azurekid for the huge amount of extremely high quality work he's done so far. We look forward to continuing and growing our involvement here to ensure that people have the best possible experience with this Solution from day one and beyond.

**Comment by v-prasadboke** [comment]
> Hi @v-prasadboke
> 
> I wanted to chime in here to introduce myself. I am a Solutions Architect from 1Password and have been working with @azurekid as he has been building this Solution.
> 
> I am wondering what the best way for me (on behalf of 1Password) to contribute to this PR so we can assist with the final polish and presentation without having to burden @azurekid with lots of little changes and enhancements?
> 
> We are excited to continue supporting this effort, and are immensely grateful to @azurekid for the huge amount of extremely high quality work he's done so far. We look forward to continuing and growing our involvement here to ensure that people have the best possible experience with this Solution from day one and beyond.

Hello @scottisloud, 
We would get this PR merged early as possible. If needed I'll commit some necessity changes that loosen the burden on azurekid

**Comment by scottisloud** [comment]
> > Hi @v-prasadboke
> > I wanted to chime in here to introduce myself. I am a Solutions Architect from 1Password and have been working with @azurekid as he has been building this Solution.
> > I am wondering what the best way for me (on behalf of 1Password) to contribute to this PR so we can assist with the final polish and presentation without having to burden @azurekid with lots of little changes and enhancements?
> > We are excited to continue supporting this effort, and are immensely grateful to @azurekid for the huge amount of extremely high quality work he's done so far. We look forward to continuing and growing our involvement here to ensure that people have the best possible experience with this Solution from day one and beyond.
> 
> Hello @scottisloud, We would get this PR merged early as possible. If needed I'll commit some necessity changes that loosen the burden on azurekid

Hi @v-prasadboke ah, okay. I think to keep things cleaner in this PR and prevent this from needing to be merged in it's in-progress state, I will work with @azurekid on a separate branch so he can bring any changes to this existing PR. We're coordinating that collaboration through a side-channel at the moment. 

Thanks for your support here @v-prasadboke!

**Comment by azurekid** [comment]
Hey All,

Previous errors are fixed, please advice on how to proceed.

**Comment by azurekid** [comment]
Changed PR to DRAFT until issues are solved

**Comment by scottisloud** [comment]
Looks like by fixing a typo I may have broken a reference somewhere, or I introduced a new typo that re-broke the pipeline. I'll see if I can track down the issues and document the fix here. If it's just a couple tweaks it may be easier for @azurekid to quickly make commit them directly here, otherwise I'll make the fix on a fork and get it merged here.

**Comment by azurekid** [comment]
> Looks like by fixing a typo I may have broken a reference somewhere, or I introduced a new typo that re-broke the pipeline. I'll see if I can track down the issues and document the fix here. If it's just a couple tweaks it may be easier for @azurekid to quickly make commit them directly here, otherwise I'll make the fix on a fork and get it merged here. 

I will pick this up @scottisloud 👍

**Comment by azurekid** [comment]
@v-prasadboke I have seemed to remove all issues, but it fails on a document link step.
Can you please take a look at this, I am not able to find the locale link it is referring to.

**Comment by v-prasadboke** [comment]
Hello @azurekid, I'll take a look at it.

File: Solutions/1Password/README.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+- Contributor role with User Access Administrator role on the Microsoft Sentinel Resource Group <br>
+**or**
+- Owner on the Microsoft Sentinel Resource Group 
+
+## Automated Installation
+
+Installing the 1Password Solution for Microsoft Sentinel is easy and can be completed in only a few minutes. 
+Just click the button below to get started with the deployment wizard. <br>
+
+[![Deploy To Azure](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.svg?sanitize=true)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2Fazurekid%2FAzure-Sentinel%2Ffeature%2F1password%2FSolutions%2F1Password%2FData%20Connectors%2F1Password%2Fazuredeploy_1Password_API_FunctionApp.json/createUIDefinitionUri/https%3A%2F%2Fraw.githubusercontent.com%2Fazurekid%2FAzure-Sentinel%2Ffeature%2F1password%2FSolutions%2F1Password%2FData%20Connectors%2F1Password%2Fdeployment%2FUiDefinition.json)
```
Comment (v-prasadboke) [inline Solutions/1Password/README.md:24]: The link used in this is from your repo. It should consist Master branch's address

**Comment by v-prasadboke** [comment]
Create a custom table at .script/tests/KqlvalidationsTests/CustomTables with name OnePasswordEventLogs_CL
You can refer to other table from the folder for more clarification. 

Also please do share sample data to test the content of the solution

File: Solutions/1Password/README.md (hunk 1 above)
Comment (azurekid) [reply Solutions/1Password/README.md]: fixed, but still an issue in the pipeline regarding links. please advice

Comment (azurekid) [reply Solutions/1Password/README.md]: replaced the link to reference azure/azure-sentinel/master/...

**Comment by azurekid** [comment]
> Create a custom table at .script/tests/KqlvalidationsTests/CustomTables with name OnePasswordEventLogs_CL You can refer to other table from the folder for more clarification.
> 
> Also please do share sample data to test the content of the solution

done

**Comment by v-prasadboke** [comment]
Hello @azurekid, Can you provide write access to your branch. Unable to pull and push commits

**Comment by v-prasadboke** [comment]
Hello @azurekid, waiting for your response

**Comment by azurekid** [comment]
> [v-prasadboke](/v-prasadboke)

Hi, sorry for the delay as I was busy to arrange stuff for the MVP Summit.
Just added your account as a contributor. Thanks for the support already ;-)

**Comment by v-prasadboke** [comment]
Hello @azurekid, working on KQL validation error. 
Will get back to you by 13 March, 2024

**Comment by v-prasadboke** [comment]
Hello @azurekid, we have one PR for OnePassword. Please check it once #9786

**Comment by scottisloud** [comment]
> Hello @azurekid, we have one PR for OnePassword. Please check it once #9786

We appreciate that others are eager to develop their own integrations between 1Password and Sentinel.

However, speaking as a 1Password employee and the person who has worked directly with @azurekid, it is 1Password's view that this is _the_ canonical 1Password Sentinel solution. As such, 1Password continues to be committed to the development of this solution, which we view (without making a formal or legal claim) as an partnership between 1Password and @azurekid.

**Comment by v-prasadboke** [comment]
Thanks @scottisloud, Will discuss about these 2 PR's with the team. And I'll update you about the same.

**Comment by scottisloud** [comment]
Hi @v-prasadboke I just wanted to check in and see if there was anything I could do assist here. It looks like there are two things standing in the way of a merge:
* How Microsoft wants to handle the other existing PR 
* The failed checks in the workflow. 

If there's any way I can help move past either of these, please let me know.

**Comment by scottisloud** [comment]
Hi @v-prasadboke. I am wondering what we need to do to move this forward. Is there a reason this was reverted to a draft and has stalled out for 2 weeks? 

Is there anything we at 1Password (A current Microsoft Cloud Partner program member) can do to help move this forward?

**Comment by v-prasadboke** [comment]
Hello @azurekid & @scottisloud, 
Sorry for the delay and we have Received green flag from Engineering team for this PR. Moving this PR to ready for review state. 

Continuing to investigate the Solution.  

- Prasad

**Comment by azurekid** [comment]
That is fantastic! Thank you so much for your support guys 🙏🏼

**Comment by v-prasadboke** [comment]
Hello @azurekid, Can you update the branch from master. 
Unfortunately the checks has stopped running.

**Comment by azurekid** [comment]
> Hello @azurekid, Can you update the branch from master. 
> Unfortunately the checks has stopped running. 

Yes, will do in about 1 hour. No access to my laptop now fue to updates.

**Comment by v-prasadboke** [comment]
Hello @azurekid, 
Can you update the branch from master once again. Sorry for asking this but pipelines had some issue and it was resolved this morning. And I'm unable to update this branch from master due to restricted access.

**Comment by v-prasadboke** [comment]
Hello @azurekid, Looks like you are committing some changes to the PR. Please let me know once its done

**Comment by v-prasadboke** [comment]
Hello @azurekid, Can you update the branch from master. I'm unable to pull latest changes to my local.

**Comment by azurekid** [comment]
> Hello @azurekid, Can you update the branch from master. I'm unable to pull latest changes to my local.

Yes, will do.
I have created some extra analytics rules and updated the workbook, so will add this to the solution package, run a validation test and update the `master`

Keep you aligned ;-)

**Comment by v-prasadboke** [comment]
Ok Thanks for the update @azurekid.

**Comment by scottisloud** [comment]
Since there's been a lot of commits recently, just checking if either @azurekid or @v-prasadboke has already updated, or is able to update, the publisher ID to be `1password1617200969773`?

**Comment by v-prasadboke** [comment]
Hey @scottisloud, I was gonna do it but I'm unable to pull latest changes to my local. 

And force pushing my changes to this branch would revert azurekids commits. 

@azurekid can you update the publisher id please

**Comment by azurekid** [comment]
> 1password1617200969773

Hi @scottisloud 
I have added the 1Password publisher ID to the solution metadata file, so this should cover it.

**Comment by azurekid** [comment]
@v-prasadboke @v-atulyadav

It took some time because I needed to repack the solution after updating the publisherId and adding some new detection rules. Especially the `Create-Sentinel-Solution V3` package creator is very lame as it adds values to the `mainTemplate.json` and `CreateUIDefinition.json` that are invalid and needs manual fixing afterwards which takes a lot of time and troubleshooting.

All files are now fixed, but it seems that the Azure-Sentinel pipeline is broken after an update see [link](https://dev.azure.com/azure/Azure-Sentinel/_build/results?buildId=74149&view=logs&j=2bbe2297-1b5b-54f3-6060-67e7b289f6c3&t=293f4a23-e36d-5efa-3a4e-ea46d4ab6d83)

image

please advice

**Comment by v-prasadboke** [comment]
update the branch from master once

**Comment by azurekid** [comment]
> update the branch from master once

done

**Comment by v-prasadboke** [comment]
Hello @azurekid, 
Please add workbook metadata to this file https://github.com/Azure/Azure-Sentinel/blob/master/Workbooks/WorkbooksMetadata.json

**Comment by azurekid** [comment]
> Hello @azurekid, Please add workbook metadata to this file https://github.com/Azure/Azure-Sentinel/blob/master/Workbooks/WorkbooksMetadata.json

Done!

**Comment by scottisloud** [comment]
@v-prasadboke Looks like we're all squared away and just awaiting review from MS. Let me or @azurekid know if there are any other changes required.
If you have an ETA for when you may be able to review this PR, then that'd be appreciated as we coordinate resources on the 1Password side of things.

**Comment by azurekid** [comment]
@v-prasadboke , @v-atulyadav, @urajat, could one of you kindly provide a response? We missed our go-live date, and it's crucial to know if there are any updates or impediments. This PR has been ongoing for 90 days now, and clarity would be greatly appreciated.

**Comment by v-prasadboke** [comment]
Hello @azurekid, Sorry was on leave till yesterday from last saturday. Was waiting for your committs to complete. 
I'll get starting with this

**Comment by azurekid** [comment]
> Hello @azurekid, Sorry was on leave till yesterday from last saturday. Was waiting for your committs to complete. 
> I'll get starting with this 
> 

Hi @v-prasadboke ,
I hope you enjoyed your holiday. It is always good to recharge the internal battery once in a while. 

Let's see if we can wrap this up together today. 
I will make sure I have availability to swiftly pickup any change request. 

@scottisloud FYI

**Comment by v-prasadboke** [comment]
Hello @azurekid, Can you update this branch from master once

**Comment by azurekid** [comment]
> Hello @azurekid, Can you update this branch from master once

Just did 👍

**Comment by azurekid** [comment]
> > Hello @azurekid, Can you update this branch from master once
> 
> Just did 👍

Hi @v-prasadboke,
I noticed you did some updates on the parser file.  
I have corrected an error in the parser as the `TimeGenerated` field was duplicated.  
Also added a comment that the parser is still in development, so no action is required on this file unless the KQLValidation pipeline has issues with the file.  

Also added some extra error handling in the function code and zip file.

**Comment by azurekid** [comment]
> Hello @azurekid, I'm trying to resolve the KQL validation issue. It is for Data connector. I've raised a Draft PR where im working on it. 
> 
> https://github.com/Azure/Azure-Sentinel/pull/10475

Okay cool!
When running the queries in the LA it comes with result.
Also when using the data connector valid results are shown in the MS Sentinel data connector

**Comment by v-prasadboke** [comment]
Hello @azurekid, On my test branch, I was looking for the cause for validation failure. 
As per my analysis and investigation I've came to known the issue which is causing the validation failure is not about the code in the file. 

Even I tried to replace the code in the data connector with another data connector. 

We can skip this validation failure but need to discuss this within team whether the validation failure is legit or not OR can it be skipped 

Thanks and sorry for the delay in response.

**Comment by azurekid** [comment]
> Hello @azurekid, On my test branch, I was looking for the cause for validation failure. As per my analysis and investigation I've came to known the issue which is causing the validation failure is not about the code in the file.
> 
> Even I tried to replace the code in the data connector with another data connector.
> 
> We can skip this validation failure but need to discuss this within team whether the validation failure is legit or not OR can it be skipped
> 
> Thanks and sorry for the delay in response.

Thanks! I am glad we came to the same conclusion regarding this.
We wanted to go live around RSA with this solution as a lot of customers are waiting for a supported integration between 1Password and Microsoft Sentinel.

Would love to see how we can put a bit more pressure on it so organizations can start to test the preview version and provide us with valuable feedback.

@scottisloud for visibility

**Comment by v-prasadboke** [comment]
> > Hello @azurekid, On my test branch, I was looking for the cause for validation failure. As per my analysis and investigation I've came to known the issue which is causing the validation failure is not about the code in the file.
> > Even I tried to replace the code in the data connector with another data connector.
> > We can skip this validation failure but need to discuss this within team whether the validation failure is legit or not OR can it be skipped
> > Thanks and sorry for the delay in response.
> 
> Thanks! I am glad we came to the same conclusion regarding this. We wanted to go live around RSA with this solution as a lot of customers are waiting for a supported integration between 1Password and Microsoft Sentinel.
> 
> Would love to see how we can put a bit more pressure on it so organizations can start to test the preview version and provide us with valuable feedback.
> 
> @scottisloud for visibility

Noted @azurekid, Will get this complete soonest

**Comment by scottisloud** [comment]
Thanks @azurekid and @v-prasadboke for all of your hard work to help bring this solution across the finish line.

**Comment by azurekid** [comment]
> > > Hello @azurekid, On my test branch, I was looking for the cause for validation failure. As per my analysis and investigation I've came to known the issue which is causing the validation failure is not about the code in the file.
> > > Even I tried to replace the code in the data connector with another data connector.
> > > We can skip this validation failure but need to discuss this within team whether the validation failure is legit or not OR can it be skipped
> > > Thanks and sorry for the delay in response.
> > 
> > Thanks! I am glad we came to the same conclusion regarding this. We wanted to go live around RSA with this solution as a lot of customers are waiting for a supported integration between 1Password and Microsoft Sentinel.
> > 
> > Would love to see how we can put a bit more pressure on it so organizations can start to test the preview version and provide us with valuable feedback.
> > 
> > @scottisloud for visibility
> 
> Noted @azurekid, Will get this complete soonest 
> 


Hey, the blog post has been published on Medium, LinkedIn, and a few security communities. People are excited to start testing, but the link in the articles still directs to my personal GitHub.

It's been 10 days now, and I think it's time to move forward. Can we escalate this PR? It's been open for months.

**Comment by v-prasadboke** [comment]
Hello, @azurekid & @scottisloud.

You know I have been working on this PR, there is a KQL validation thats throwing an error for null value.

But, while investigating the error I did not notice any missing or null values.

I submitted a test PR and have been working on it
And as you said the data connector is working and ingests logs but I dont want these KQL validation error to be a part of any trouble in the future even if we skip it. 

I'm in connect with Internal resources regarding this KQL failure. 

I apologize @azurekid and @scottisloud for the inconvenience and delay with this PR.

**Comment by scottisloud** [comment]
Hi @v-prasadboke thanks for the update. I see it's been about a week since #10536 had commits and two days since your post above. 

We agree we'd prefer to not have a validation error cause trouble in the future, however the ongoing delays here are holding up our ability to deliver a feature to our mutual customers. 

Do we know the validation failures flagged by your GH workflow are meaningful failures? In our testing we've seen no actual impacts to the behaviour of the solution, despite the errors your workflow flags. 

We have had to reschedule formal release and marketing a couple times already, and we'd like to avoid further last-minute delays. To that end, is there a chance you could provide us with an approximate ETA for a resolution to the GH workflow issues? That would allow us to anticipate and schedule the release in a way that accommodates your ongoing efforts here.

**Comment by v-prasadboke** [comment]
Hello @azurekid & @azurekid, The KQL validation is resolved. The function app is new to me. Can we get on a call tomorrow so that I can get an understanding of this function apps working and get this PR merged by tomorrow EOD. 

I work in IST time zone and my mail id is v-prasadboke@microsoft.com  
You can ping me on teams so that schedule about the same.

**Comment by azurekid** [comment]
Thank you very much!
I have sent a teams invite for tomorrow 👍

**Comment by v-prasadboke** [comment]
image

Hello @azurekid, Please follow this guidelines for offerid 
https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/README.md

**Comment by azurekid** [comment]
> image
> 
> Hello @azurekid, Please follow this guidelines for offerid https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/README.md

just pushed the change to the `solutionMetadata` file

**Comment by v-prasadboke** [comment]
@azurekid maybe you missed on committing the change requested for OfferID. Can you confirm on this once.

**Comment by azurekid** [comment]
> @azurekid maybe you missed on committing the change requested for OfferID. Can you confirm on this once.

Sorry, pushed it to the wrong branch.
Updated the code in correct branch now

**Review by v-prasadboke — APPROVED** [review]

**Pull request merged**